### PR TITLE
Kl/chain balance fix

### PR DESCRIPTION
**Pull request by kelemeno**
# What ❔

## Checklist

- [ ] PR title corresponds to the body of PR (we generate changelog entries from PRs).
- [ ] Tests for the changes have been added / updated.
- [ ] Documentation comments have been added / updated.

### CODE REVIEW

File: l1-contracts/contracts/common/libraries/DataEncoding.sol [hunk 1]
```diff
@@ -117,4 +117,32 @@ library DataEncoding {
             revert UnsupportedEncodingVersion();
         }
     }
+
+    function decodeTokenData(
+        bytes calldata _tokenData
+    ) internal pure returns (uint256 chainId, bytes memory name, bytes memory symbol, bytes memory decimals) {
+        bytes1 encodingVersion = _tokenData[0];
+        // kl todo check correct
+        if (encodingVersion == LEGACY_ENCODING_VERSION) {
```
Comment (kelemeno) [inline l1-contracts/contracts/common/libraries/DataEncoding.sol:126]: Will this work Stas?

File: l1-contracts/contracts/bridge/BridgedStandardERC20.sol [hunk 2]
```diff
@@ -37,7 +38,7 @@ contract BridgedStandardERC20 is ERC20PermitUpgradeable, IBridgedStandardToken,
     address public override l2Bridge;
 
     /// @dev Address of the L1 token that can be deposited to mint this L2 token
-    address public override l1Address;
+    address public override originToken;
```
Comment (StanislavBreadless) [inline l1-contracts/contracts/bridge/BridgedStandardERC20.sol:41]: probably better to update the comment as well

File: l1-contracts/contracts/bridge/ntv/L1NativeTokenVault.sol [hunk 3]
```diff
@@ -37,6 +37,11 @@ contract L1NativeTokenVault is IL1NativeTokenVault, IL1AssetHandler, NativeToken
     /// @dev Era's chainID
     uint256 public immutable ERA_CHAIN_ID;
 
+    /// @dev Maps token balances for each chain to prevent unauthorized spending across ZK chains.
+    /// This serves as a security measure until hyperbridging is implemented.
+    /// NOTE: this function may be removed in the future, don't rely on it!
+    mapping(uint256 chainId => mapping(bytes32 token => uint256 balance)) public chainBalance;
```
Comment (StanislavBreadless) [inline l1-contracts/contracts/bridge/ntv/L1NativeTokenVault.sol:43]:
```suggestion
    mapping(uint256 chainId => mapping(bytes32 assetId => uint256 balance)) public chainBalance;
```

this way it is more consistent with existing naming

File: l1-contracts/contracts/bridge/ntv/L1NativeTokenVault.sol [hunk 4]
```diff
@@ -224,14 +235,48 @@ contract L1NativeTokenVault is IL1NativeTokenVault, IL1AssetHandler, NativeToken
         }
     }
 
+    function _deployBridgedToken(
+        uint256 _originChainId,
+        address _originToken,
+        bytes memory _erc20Data
+    ) internal virtual override returns (address, uint256) {
+        originChainId[DataEncoding.encodeNTVAssetId(_originChainId, _originToken)] = _originChainId;
```
Comment (StanislavBreadless) [inline l1-contracts/contracts/bridge/ntv/L1NativeTokenVault.sol:243]: is it the same as on L2 ntv?

Comment (kelemeno) [reply l1-contracts/contracts/bridge/ntv/L1NativeTokenVault.sol:243]: No, the _originChainId is the origin of the message, which might not be the origin of the token
On the l2 we get the tokenOriginChainId returned from the bridgeInit

File: l1-contracts/contracts/bridge/ntv/NativeTokenVault.sol [hunk 5]
```diff
@@ -111,7 +110,7 @@ abstract contract NativeTokenVault is INativeTokenVault, IAssetHandler, Ownable2
     ) external payable override onlyAssetRouter whenNotPaused {
         address receiver;
         uint256 amount;
-        if (isTokenNative[_assetId]) {
+        if (originChainId[_assetId] == block.chainid) {
```
Comment (StanislavBreadless) [inline l1-contracts/contracts/bridge/ntv/NativeTokenVault.sol:113]: please add a comment that explains why is it safe even if the token has been bridged prior to this update (I suspect, it is an indirect check when we deploy the token)

File: l1-contracts/contracts/bridge/ntv/NativeTokenVault.sol [hunk 6]
```diff
@@ -43,16 +43,14 @@ abstract contract NativeTokenVault is INativeTokenVault, IAssetHandler, Ownable2
...
+    /// @dev A mapping assetId => tokenAddress
+    mapping(bytes32 assetId => uint256 chainId) public originChainId;
 
     /// @dev A mapping assetId => tokenAddress
     mapping(bytes32 assetId => address tokenAddress) public tokenAddress;
 
-    /// @dev A mapping assetId => isTokenBridged
-    mapping(bytes32 assetId => bool bridged) public isTokenNative;
+    // /// @dev A mapping assetId => isTokenBridged
+    // mapping(bytes32 assetId => bool bridged) public isTokenNative;
```
Comment (StanislavBreadless) [inline l1-contracts/contracts/bridge/ntv/NativeTokenVault.sol:53]: ?

File: l1-contracts/contracts/bridge/ntv/NativeTokenVault.sol [hunk 7]
```diff
@@ -208,7 +204,17 @@ abstract contract NativeTokenVault is INativeTokenVault, IAssetHandler, Ownable2
             receiver: _receiver,
             amount: _amount
         });
-        _bridgeMintData = _data;
+        bytes memory erc20Metadata;
+        {
+            erc20Metadata = getERC20Getters(bridgedToken, true, originChainId[_assetId]);
```
Comment (StanislavBreadless) [inline l1-contracts/contracts/bridge/ntv/NativeTokenVault.sol:209]: let's add a comment that explains why `originChainId` is non-zero here

Comment (StanislavBreadless) [reply l1-contracts/contracts/bridge/ntv/NativeTokenVault.sol:209]: maybe even better, we could have an explicit error if either the address of the origin chain id is 0

File: l1-contracts/contracts/common/libraries/DataEncoding.sol (hunk 1 above)
Comment (StanislavBreadless) [reply l1-contracts/contracts/common/libraries/DataEncoding.sol:126]: LGTM, though maybe a different constant variable should be introduced

File: l1-contracts/contracts/bridge/ntv/INativeTokenVault.sol [hunk 8]
```diff
@@ -17,20 +17,23 @@ interface INativeTokenVault {
     /// @notice The AssetRouter contract
     function ASSET_ROUTER() external view returns (IAssetRouterBase);
 
-    /// @notice Returns the total number of specific tokens locked for some chain
-    function chainBalance(uint256 _chainId, bytes32 _assetId) external view returns (uint256);
-
     /// @notice Returns if the bridged version of bridged token has been deployed
-    function isTokenNative(bytes32 assetId) external view returns (bool);
+    // function isTokenNative(bytes32 assetId) external view returns (bool);
```
Comment (StanislavBreadless) [inline l1-contracts/contracts/bridge/ntv/INativeTokenVault.sol:21]: commented out code

File: l1-contracts/contracts/bridge/ntv/INativeTokenVault.sol [hunk 9]
```diff
@@ -17,20 +17,23 @@ interface INativeTokenVault {
...
+    /// @notice Allows the bridge to register a token address for the vault.
+    /// @notice No access control is ok, since the bridging of tokens should be permissionless. This requires permissionless registration.
     function registerToken(address _l1Token) external;
 
     /// @notice Used to get the assetId of a token
     function getAssetId(uint256 _chainId, address _tokenAddress) external view returns (bytes32);
 
     /// @notice Used to get the the ERC20 data for a token
-    function getERC20Getters(address _token) external view returns (bytes memory);
+    function getERC20Getters(address _token, bool _legacy, uint256 _originChainId) external view returns (bytes memory);
```
Comment (StanislavBreadless) [inline l1-contracts/contracts/bridge/ntv/INativeTokenVault.sol:36]: maybe I am missing something, but it seems like in all places where this function is used, _legacy = true

File: l1-contracts/contracts/bridge/ntv/L1NativeTokenVault.sol [hunk 10]
```diff
@@ -113,8 +118,14 @@ contract L1NativeTokenVault is IL1NativeTokenVault, IL1AssetHandler, NativeToken
         L1_NULLIFIER.nullifyChainBalanceByNTV(_targetChainId, _token);
     }
 
+    /// @inheritdoc INativeTokenVault
+    function registerToken(address _nativeToken) external override(INativeTokenVault, NativeTokenVault) {
+        originChainId[DataEncoding.encodeNTVAssetId(block.chainid, _nativeToken)] = block.chainid;
```
Comment (StanislavBreadless) [inline l1-contracts/contracts/bridge/ntv/L1NativeTokenVault.sol:123]: I am not sure if its behavior is any different from the parent function

File: l1-contracts/contracts/bridge/ntv/L2NativeTokenVault.sol [hunk 11]
```diff
@@ -128,6 +128,20 @@ contract L2NativeTokenVault is IL2NativeTokenVault, NativeTokenVault {
...
+        uint256 _originChainId,
+        address _originToken,
+        bytes memory _erc20Data
+    ) internal virtual override returns (address, uint256) {
+        (address deployedToken, uint256 tokenOriginChainId) = super._deployBridgedToken(
+            _originChainId,
+            _originToken,
+            _erc20Data
+        );
+        originChainId[DataEncoding.encodeNTVAssetId(tokenOriginChainId, _originToken)] = tokenOriginChainId;
```
Comment (StanislavBreadless) [inline l1-contracts/contracts/bridge/ntv/L2NativeTokenVault.sol:141]: seems same as parent

File: l1-contracts/contracts/bridge/ntv/INativeTokenVault.sol (hunk 9 above)
Comment (kelemeno) [reply l1-contracts/contracts/bridge/ntv/INativeTokenVault.sol:32]: Ok, then I did not do it properly, the idea is to have

**Pull request merged**